zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/MultipleAddresses.java

```diff
@@ -0,0 +1,187 @@
+        this.addresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        this.addresses.addAll(addresses);
+        this.timeout = timeout;
+    }
+
+    public MultipleAddresses(InetSocketAddress address,  int timeout) {
+        addresses = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        addresses.add(address);
+        this.timeout = timeout;
+    }
+
+    public int getTimeout() {
+        return timeout;
+    }
+
+    public void setTimeout(int timeout) {
+        this.timeout = timeout;
+    }
+
+    public boolean isEmpty() {
+        return addresses.isEmpty();
+    }
+
+    /**
+     * Returns all addresses.
+     *
+     * @return list of all InetSocketAddress
+     */
+    public List<InetSocketAddress> getAllAddresses() {
+        return new LinkedList<>(addresses);
+    }
+
+    /**
+     * Returns wildcard addresses for all ports
+     *
+     * @return list of InetSocketAddress with wildcards for all ports
+     */
+    public List<InetSocketAddress> getWildcardAddresses() {
+       return addresses.stream().map(a -> new InetSocketAddress(a.getPort())).distinct().collect(Collectors.toList());
+    }
+
+    /**
+     * Returns all ports
+     *
+     * @return list of all ports
+     */
+    public List<Integer> getAllPorts() {
+        return addresses.stream().map(InetSocketAddress::getPort).distinct().collect(Collectors.toList());
+    }
+
+    public void addAddress(InetSocketAddress address) {
+        addresses.add(address);
+    }
+
+    /**
+     * Returns reachable address. If none is reachable than throws exception.
+     *
+     * @return address which is reachable.
+     * @throws NoRouteToHostException if none address is reachable
+     */
+    public InetSocketAddress getValidAddress() throws NoRouteToHostException {
+        AtomicReference<InetSocketAddress> address = new AtomicReference<>(null);
+        getInetSocketAddressStream().forEach(addr -> checkIfAddressIsReachableAndSet(addr, address));
+
+        if(address.get() != null)
+            return address.get();
+        else
+            throw new NoRouteToHostException("No valid address among " + addresses);
+    }
+
+    /**
+     * Performs a DNS lookup for addresses.
+     *
+     * If the DNS lookup fails, than address remain unmodified.
+     */
+    public void recreateSocketAddresses() {
+        Set<InetSocketAddress> temp = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        temp.addAll(getInetSocketAddressStream().map(this::recreateSocketAddress).collect(Collectors.toSet()));
+        addresses = temp;
+    }
+
+    public InetSocketAddress getOne() {
+        return addresses.iterator().next();
+    }
+
+    private void checkIfAddressIsReachableAndSet(InetSocketAddress address,
+                                                 AtomicReference<InetSocketAddress> reachableAddress) {
+        for(int i = 0; i < 5 && reachableAddress.get() == null; i++) {
```

Comment:
It might be a bit overkill to probe all server addresses on every connect request. This method should quit immediately once `reachableAddress` is set.